### PR TITLE
Fixes 1579: update url on remote if url changed

### PR DESCRIPTION
**Pull request by jlsherrill**
## Summary

Supports updating url of a repository with snapshotting enabled.
Will update the remote to the new url as part of the snapshot
If the url is being updated, and a snapshot is in progress, we reject the request

## Testing steps

make this request:
```
POST http://localhost:8000/api/content-sources/v1.0/repositories/
Content-Type: application/json
x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiJqZG9lIn0sImludGVybmFsIjp7Im9yZ19pZCI6IjEyMyJ9fX0K

{"name":"sample zoo","url":"https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/","snapshot": true}
```

Then update the repo (change the uuid to match what is returned):

```
PATCH http://localhost:8000/api/content-sources/v1.0/repositories/1538f16f-324e-4d8e-95d4-483d73b3cebd
Content-Type: application/json
x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiJqZG9lIn0sImludGVybmFsIjp7Im9yZ19pZCI6IjEyMyJ9fX0K
```

Then check the snapshots table:

```
make db-cli-connect
select * from snapshots
```

you should see 2 created snapshots now.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-1579

**Comment by swadeley** [comment]
Hi

Would changing `Restrict architecture` value be enough to trigger a new snapshot?

In which pod would the `make db-cli-connect` command work?

Thank you

File: pkg/handler/repositories.go [hunk 1]
```diff
@@ -299,11 +299,31 @@ func (rh *RepositoryHandler) update(c echo.Context, fillDefaults bool) error {
...
+		return ce.NewErrorResponse(ce.HttpCodeForDaoError(err), "Error fetching repository", err.Error())
+	}
+
+	if repoParams.URL != nil && repoConfig.URL != *repoParams.URL {
+		snapInProgress, err := rh.DaoRegistry.TaskInfo.IsSnapshotInProgress(orgID, uuid)
+		if err != nil {
+			return ce.NewErrorResponse(ce.HttpCodeForDaoError(err), "Error checking if snapshot is in progress", err.Error())
+		}
+		if snapInProgress {
+			return ce.NewErrorResponse(http.StatusBadRequest, "Cannot delete repository while snapshot is in progress", "")
```
Comment (rverdile) [inline pkg/handler/repositories.go:314]:
```suggestion
			return ce.NewErrorResponse(http.StatusBadRequest, "Cannot update repository while snapshot is in progress", "")
```

File: pkg/handler/repositories.go [hunk 2]
```diff
@@ -299,11 +299,31 @@ func (rh *RepositoryHandler) update(c echo.Context, fillDefaults bool) error {
...
 	}
-	if err := rh.DaoRegistry.RepositoryConfig.Update(orgID, uuid, repoParams); err != nil {
+
+	repoConfig, err := rh.DaoRegistry.RepositoryConfig.Fetch(orgID, uuid)
+	if err != nil {
+		return ce.NewErrorResponse(ce.HttpCodeForDaoError(err), "Error fetching repository", err.Error())
+	}
+
+	if repoParams.URL != nil && repoConfig.URL != *repoParams.URL {
+		snapInProgress, err := rh.DaoRegistry.TaskInfo.IsSnapshotInProgress(orgID, uuid)
```
Comment (rverdile) [inline pkg/handler/repositories.go:309]: this takes the repository UUID, not the repositoryConfig UUID, so at the moment I can change the URLs while a snapshot is being created

Comment (jlsherrill) [reply pkg/handler/repositories.go:309]: doh, this probably broke on a rebase

Comment (jlsherrill) [reply pkg/handler/repositories.go]: should be fixed now

**Review by rverdile — APPROVED** [review]
lgtm!

**Comment by swadeley** [comment]
> Hi
> 
> Would changing `Restrict architecture` value be enough to trigger a new snapshot?
> 
> In which pod would the `make db-cli-connect` command work?
> 
> Thank you

Justin Says:
Edit repo, list snapshots, see the change

**Comment by swadeley** [comment]
Hi

Created a repo, listed its snapshots:

In [9]: app.content_sources.rest_client.repositories_api.list_snapshots('80fc43a0-520c-45e0-b755-e814e3f619ad')
2023-07-05 18:31:14.070 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=v5HNU8WEjxMukwwzmVq7Ygmy6Ot7zvmZ, params=[]
Out[9]: 
{'data': [{'content_counts': {'rpm.package': 1},
           'created_at': '2023-07-05T17:26:34.957859Z',
           'distribution_path': '80fc43a0-520c-45e0-b755-e814e3f619ad/c2fbb229-af22-476d-bebd-037d050c44ba'}],
 'links': {'first': '/api/content-sources/v1/repositories/80fc43a0-520c-45e0-b755-e814e3f619ad/snapshots/?limit=100&offset=0',
           'last': '/api/content-sources/v1/repositories/80fc43a0-520c-45e0-b755-e814e3f619ad/snapshots/?limit=100&offset=0'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}


Change URL to point to new repo with different packages:

Out[18]: 
{'data': [{'content_counts': {'rpm.package': 1},
           'created_at': '2023-07-05T17:26:34.957859Z',
           'distribution_path': '80fc43a0-520c-45e0-b755-e814e3f619ad/c2fbb229-af22-476d-bebd-037d050c44ba'},
          {'content_counts': {'rpm.advisory': 4, 'rpm.package': 32},
           'created_at': '2023-07-05T17:44:57.23853Z',
           'distribution_path': '80fc43a0-520c-45e0-b755-e814e3f619ad/5259c272-6770-4e66-bfd6-5c55201224d3'}],
 'links': {'first': '/api/content-sources/v1/repositories/80fc43a0-520c-45e0-b755-e814e3f619ad/snapshots/?limit=100&offset=0',
           'last': '/api/content-sources/v1/repositories/80fc43a0-520c-45e0-b755-e814e3f619ad/snapshots/?limit=100&offset=0'},
 'meta': {'count': 2, 'limit': 100, 'offset': 0}}

There we see :  {'rpm.advisory': 4, 'rpm.package': 32},
and : {'count': 2

Thank you

**Pull request merged**